src/crypto/algorithms/megolm.js

```diff
@@ -328,31 +328,35 @@ MegolmEncryption.prototype._shareKeyWithDevices = function(session, devicesByUse
                     ciphertext: {},
                 };
 
-                olmlib.encryptMessageForDevice(
-                    encryptedContent.ciphertext,
-                    self._userId,
-                    self._deviceId,
-                    self._olmDevice,
-                    userId,
-                    deviceInfo,
-                    payload,
-                );
-
                 if (!contentMap[userId]) {
                     contentMap[userId] = {};
                 }
 
                 contentMap[userId][deviceId] = encryptedContent;
-                haveTargets = true;
+
+                promises.push(
+                    olmlib.encryptMessageForDevice(
+                        encryptedContent.ciphertext,
+                        self._userId,
+                        self._deviceId,
+                        self._olmDevice,
+                        userId,
+                        deviceInfo,
+                        payload,
+                    ),
+                );
             }
         }
 
-        if (!haveTargets) {
+        if (promises.length === 0) {
+            // no devices to send to
             return Promise.resolve();
         }
 
-        // TODO: retries
-        return self._baseApis.sendToDevice("m.room.encrypted", contentMap);
+        return Promise.all(promises).then(() => {
```

Comment:
I'm failing to understand the problem you're alluding to.

Previously, if `encryptMessageForDevice` failed on the Nth attempt, we would bail out, despite having successfully encrypted (and not sent) messages for the previous N-1 devices.

Now, if encryption attempt N fails, then again, we bail out in the same way. (`encryptMessageForDevice` might succeed for N+1 onwards, but that doesn't matter any more than 0..(N-1).) So I don't see what's changing here.

From the point of view of the Olm protocol: no it doesn't matter if a message we encrypted gets lost or never gets sent: the protocol recovers from that.